client/hwlib/src/collectors/cpuinfo.rs

```diff
@@ -0,0 +1,197 @@
+/* Copyright 2024 Canonical Ltd.
+ * All rights reserved.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Written by:
+ *        Nadzeya Hutsko <nadzeya.hutsko@canonical.com>
+ */
+
+use anyhow::{Error, Result};
+use itertools::Itertools;
+use std::collections::HashMap;
+use std::fs::{self, read_to_string};
+
+#[derive(Debug)]
+pub struct CpuInfo {
+    pub platform: String,
+    pub cores_count: usize,
+    pub cpu_type: String,
+    pub model: String,
+    pub model_number: String,
+    pub model_version: String,
+    pub model_revision: String,
+    pub cache: i64,
+    pub bogomips: i64,
+    pub speed: u64,
+}
+
+impl CpuInfo {
+    /// Parse cpuinfo file the same way it's done in checkbox:
+    /// https://github.com/canonical/checkbox/blob/3789fdd/checkbox-support/checkbox_support/parsers/cpuinfo.py
+    pub fn from_file(cpuinfo_filepath: &'static str) -> Result<CpuInfo> {
+        let mut attributes: HashMap<&str, &str> = HashMap::new();
+        let mut cores_count = 0;
+
+        let raw_cpuinfo = read_to_string(cpuinfo_filepath)?;
+
+        for line in raw_cpuinfo.lines() {
+            if line.trim().is_empty() {
+                continue;
+            }
+
+            let parts: Option<(_, _)> = line.split(':').collect_tuple();
+            if let Some((key, value)) = parts {
+                let key = key.trim();
+                let value = value.trim();
+                if key == "processor" {
+                    cores_count += 1;
+                }
+                attributes.insert(key, value);
+            }
+        }
+
+        let arch = std::env::consts::ARCH;
+        let speed_str = attributes.get("cpu MHz").copied();
+        let speed = CpuSpeed::try_from(speed_str)?.m_hz();
+
+        let platform = arch.to_string();
+        let model = attributes
+            .get("Model")
+            .or_else(|| attributes.get("model name"))
+            .unwrap_or(&arch)
+            .to_string();
+
+        let cpu_type = attributes.get("vendor_id").unwrap_or(&arch).to_string();
+        let model_number = attributes
+            .remove("cpu family")
+            .unwrap_or_default()
```

Comment:
The server DB doesn't allow NULL for string fields to avoid confusions between `NULL` and `""` being considered as different values, but in fact they just don't store any data. So, we don't use `None` or `Option` for strings, and empty strings are preferable